### PR TITLE
Add polygons mask support to COCOReader

### PR DESCRIPTION
**Pull request by Kh4L**
#### Why we need this PR?
We need the support of polygons masks (used for instance segmentation) in the existing `COCOReader`, which only supports the object detection part of [COCO-style datasets](http://cocodataset.org/#format-data). 

#### What happened in this PR?
 - Explain solution of the problem, new feature added.
Add segmentation support in the RapidJSON-based parser. 
Optional argument added to `COCOReader`, `masks` that add two additional outputs to the operator (`masks_meta` and `masks_coords`, described in the operator docstring).
 - What was changed, added, removed?
Changed: `COCOReader`, `COCOLoader` and their tests.
 - Was this PR tested? How?
Tests case added in the existing test suite and DALI_EXTRA accordingly updated for polygons masks (https://github.com/NVIDIA/DALI_extra/pull/31).
 - Were docs and examples updated, if necessary?
Op docstring explaining the internal polygons mask format in DALI.

### CODE REVIEW

**Comment by Kh4L** [comment]
> I think you need to adjust `DumpMetaFiles` and `ParseMetafiles` for this new data as well.

Done

File: dali/operators/reader/coco_reader_op.cc [hunk 1]
```diff
@@ -39,6 +39,17 @@ Tensor (`m` * `[x, y, w, h] or `m` * [left, top, right, bottom]`) and labels as
   .AddOptionalArg("ltrb",
       R"code(If true, bboxes are returned as [left, top, right, bottom], else [x, y, width, height].)code",
       false)
+  .AddOptionalArg("masks",
+      R"code(If true, segmentation masks are read and returned as polygons.
+Each mask can be one or more polygons. A polygon is a list of points (2 floats).
+For a given sample, the polygons are represented by two tensors:
+  `masks_meta` -> list of tuples (mask_idx, start_idx, count)
```
Comment (JanuszL) [inline dali/operators/reader/coco_reader_op.cc]: Maybe we should stick to sparse tensor format - I mean to return 3 tensors:
- values
- start
- count

File: dali/operators/reader/coco_reader_op.cc [hunk 2]
```diff
@@ -39,6 +39,17 @@ Tensor (`m` * `[x, y, w, h] or `m` * [left, top, right, bottom]`) and labels as
...
+  `masks_coords`-> list of (x,y) coordinates
+One mask can have one or more `masks_meta` having the same `mask_idx`, which means that the mask for that given
+index consists of several polygons).
+`start_idx` indicates the index of the first coords in `masks_coords`.
+Skip `is_crowd=1` annotations.)code",
```
Comment (JanuszL) [inline dali/operators/reader/coco_reader_op.cc]: ?

File: dali/operators/reader/coco_reader_op.cc (hunk 1 above)
Comment (Kh4L) [reply dali/operators/reader/coco_reader_op.cc]: We also need to carry `mask_idx` to know which polygon belongs to which mask

File: dali/operators/reader/coco_reader_op.cc (hunk 2 above)
Comment (Kh4L) [reply dali/operators/reader/coco_reader_op.cc]: It refers to the COCO dataset format.
```
"iscrowd": 0 if your segmentation based on polygon (object instance)
"iscrowd": 1 if your segmentation based uncompressed RLE (crowd)
```
I'll expand upon the docstring and change it to `iscrowd`

Comment (JanuszL) [reply dali/operators/reader/coco_reader_op.cc]: The more verbose description is the thing I'm asking for.

Comment (Kh4L) [reply dali/operators/reader/coco_reader_op.cc]: Done

**Comment by Kh4L** [comment]
!build

**Pull request merged**